xline/src/storage/kvstore.rs

```diff
@@ -187,55 +208,112 @@ impl KvStoreBackend {
+    fn sort_kvs(kvs: &mut [KeyValue], sort_order: SortOrder, sort_target: SortTarget) {
+        match (sort_target, sort_order) {
+            (SortTarget::Key, SortOrder::None) => {}
+            (SortTarget::Key, SortOrder::Ascend) => {
+                kvs.sort_by(|a, b| a.key.cmp(&b.key));
+            }
+            (SortTarget::Key, SortOrder::Descend) => {
+                kvs.sort_by(|a, b| b.key.cmp(&a.key));
+            }
+            (SortTarget::Version, SortOrder::Ascend | SortOrder::None) => {
+                kvs.sort_by(|a, b| a.version.cmp(&b.version));
+            }
+            (SortTarget::Version, SortOrder::Descend) => {
+                kvs.sort_by(|a, b| b.version.cmp(&a.version));
+            }
+            (SortTarget::Create, SortOrder::Ascend | SortOrder::None) => {
+                kvs.sort_by(|a, b| a.create_revision.cmp(&b.create_revision));
+            }
+            (SortTarget::Create, SortOrder::Descend) => {
+                kvs.sort_by(|a, b| b.create_revision.cmp(&a.create_revision));
+            }
+            (SortTarget::Mod, SortOrder::Ascend | SortOrder::None) => {
+                kvs.sort_by(|a, b| a.mod_revision.cmp(&b.mod_revision));
+            }
+            (SortTarget::Mod, SortOrder::Descend) => {
+                kvs.sort_by(|a, b| b.mod_revision.cmp(&a.mod_revision));
+            }
+            (SortTarget::Value, SortOrder::Ascend | SortOrder::None) => {
+                kvs.sort_by(|a, b| a.value.cmp(&b.value));
+            }
+            (SortTarget::Value, SortOrder::Descend) => {
+                kvs.sort_by(|a, b| b.value.cmp(&a.value));
+            }
+        };
+    }
+
+    /// fitler kvs by `{max,min}_{mod,create}_revision`
+    fn filter_kvs(
+        kvs: &mut Vec<KeyValue>,
+        max_mod_revision: i64,
+        min_mod_revision: i64,
+        max_create_revision: i64,
+        min_create_revision: i64,
+    ) {
+        if max_mod_revision > 0 {
+            kvs.retain(|kv| kv.mod_revision <= max_mod_revision);
+        }
+        if min_mod_revision > 0 {
+            kvs.retain(|kv| kv.mod_revision >= min_mod_revision);
+        }
+        if max_create_revision > 0 {
+            kvs.retain(|kv| kv.create_revision <= max_create_revision);
+        }
+        if min_create_revision > 0 {
+            kvs.retain(|kv| kv.create_revision >= min_create_revision);
+        }
+    }
+
     /// Handle `RangeRequest`
     fn handle_range_request(&self, req: &RangeRequest) -> RangeResponse {
-        let mut kvs = self.get_range(&req.key, &req.range_end, req.revision);
-        debug!("handle_range_request kvs {:?}", kvs);
+        debug!("handle_range_request kvs");
+        let limit = if (req.sort_order() != SortOrder::None)
+            || (req.max_mod_revision != 0)
+            || (req.min_mod_revision != 0)
+            || (req.max_create_revision != 0)
+            || (req.min_create_revision != 0)
+            || (req.limit == 0)
```

Comment:
limit is just limit. Why do we get unlimited results if we've set max_create_revision?

Reply:
because we need filter results before limit them, if we just get limited results, The final result may be less than limit